## Chapter 2 - Part 1


In [13]:
"""
Bài 1. Viết chương trình nhập vào một chuỗi
Xuất ra màn hình chuỗi ký tự hoa
Xuất ra màn hình chuỗi ký tự có ký tự đầu là ký tự hoa
Xuất ra màn hình chuỗi ký tự thường
"""

'\nBài 1. Viết chương trình nhập vào một chuỗi\nXuất ra màn hình chuỗi ký tự hoa\nXuất ra màn hình chuỗi ký tự có ký tự đầu là ký tự hoa\nXuất ra màn hình chuỗi ký tự thường\n'

In [14]:
st = input("Enter your string:")
st_upper = st.upper()
st_lower = st.lower()
st_title = st.title()

print("A Upper String is: " + st_upper)
print("A Lower String is : " + st_lower)
print("A Title String is : " + st_title)

A Upper String is: 
A Lower String is : 
A Title String is : 


In [15]:
"""
Bài 2. Cho chuỗi biểu diễn địa chỉ email:   " minhnhutvh@gmail.com"
Rút trích và hiển thị chuỗi "gmail.com" (Đây chính là  tên Host)
"""

'\nBài 2. Cho chuỗi biểu diễn địa chỉ email:   " minhnhutvh@gmail.com"\nRút trích và hiển thị chuỗi "gmail.com" (Đây chính là  tên Host)\n'

In [16]:
data = "minhnhutvh@gmail.com"
position = data.find("@")
host = data[position + 1 :]
print(host)

gmail.com


In [17]:
"""
Bài 3. Cho chuỗi biểu diễn:
        ‘minhnhutvh@gmai.com Sat Jan 5 09:14:16’
Rút trích và hiển thị chuỗi “gmail.com”
"""

'\nBài 3. Cho chuỗi biểu diễn:\n        ‘minhnhutvh@gmai.com Sat Jan 5 09:14:16’\nRút trích và hiển thị chuỗi “gmail.com”\n'

In [18]:
data = "minhnhutvh@gmail.com Sat Jan 5 09:14:16"
Start_position = data.find("@")  #  10
End_position = data.find(" ", Start_position)  #  26
host = data[Start_position + 1 : End_position]
print(host)

gmail.com


In [19]:
"""Những bài thực hành còn lại. Sinh viên tự làm"""

'Những bài thực hành còn lại. Sinh viên tự làm'

In [20]:
# Ex4

st1 = input("Enter your string: ")
st2 = input(" Enter the word that needs to replace: ")
st3 = input(" Enter the word after being replaced: ")

st1 = st1.replace(st2, st3)
print(st1)

In [21]:
# Ex5

st = input("Enter your string: ")
First_five_characters = st[:5]
Last_five_characters = st[len(st) - 5 :]
Str_1_line = 4 * (st + " ")
Str_4_line = 4 * (st + "\n")

print("First five characters are: " + First_five_characters)
print("Last five characters are: " + Last_five_characters)
print("Four strings of one line are: " + Str_1_line)
print("Four strings of four line are: \n" + Str_4_line)

First five characters are: 
Last five characters are: 
Four strings of one line are:     
Four strings of four line are: 







In [23]:
# Ex6

from math import pi, pow
from utils.formatInput import floatInput

r, h = floatInput("Enter r:"), floatInput("Enter h:")
V = pi * pow(r, 2) * h

print(f"V(r={r}, h={h}) =", V)

V(r=10.0, h=12.0) = 3769.9111843077517


In [24]:
# Ex7

from math import sin, degrees, pi
from utils.formatInput import floatInput as fI


def isValidTriangle(
    edge_a: float,
    edge_b: float,
    edge_c: float,
    angle_A: float,
    angle_B: float,
    angle_C: float,
):
    edge_greater_then_zero = edge_a > 0 and edge_b > 0 and edge_c > 0
    print("edge_greater_then_zero: ", edge_greater_then_zero)

    angle_in_valid_range = 0 < angle_A < pi and 0 < angle_B < pi and 0 < angle_C < pi
    print("angle_in_valid_range: ", angle_in_valid_range)

    # góc input đơn vị rad, dùng degrees thành float có thể bị lệch khi so trực tiếp với 180, so sánh hiệu với float nhỏ
    total_angle_is_180 = (
        abs(degrees(angle_A) + degrees(angle_B) + degrees(angle_C) - 180) < 0.0000001
    )
    print("total_angle_is_180: ", total_angle_is_180, "\n")

    return edge_greater_then_zero and angle_in_valid_range and total_angle_is_180


def triangleS(edge_1: float, edge_2: float, mid_angle: float):
    result = 1 / 2 * edge_1 * edge_2 * sin(mid_angle)
    print(
        f"triangleS(edge_1={edge_1}, edge_2={edge_2}, mid_angle={mid_angle}) =", result
    )
    return result


"""
Test
a: 2
b: 2
c: 2
A: 1.0471975512
B: 1.0471975512
C: 1.0471975512
"""
edge_a, edge_b, edge_c = fI("edge a:"), fI("edge b:"), fI("edge c:")
angle_A, angle_B, angle_C = (
    fI("angle A (rad):"),
    fI("angle B (rad):"),
    fI("angle C (rad):"),
)

if not isValidTriangle(edge_a, edge_b, edge_c, angle_A, angle_B, angle_C):
    raise ValueError("Invalid triangle")

triangleS(edge_a, edge_b, angle_C)
triangleS(edge_a, edge_c, angle_B)
triangleS(edge_b, edge_c, angle_A)

edge_greater_then_zero:  True
angle_in_valid_range:  True
total_angle_is_180:  True 

triangleS(edge_1=2.0, edge_2=2.0, mid_angle=1.0471975512) = 1.7320508075722796
triangleS(edge_1=2.0, edge_2=2.0, mid_angle=1.0471975512) = 1.7320508075722796
triangleS(edge_1=2.0, edge_2=2.0, mid_angle=1.0471975512) = 1.7320508075722796


1.7320508075722796

In [25]:
# Ex8

from math import sqrt
from utils.formatInput import floatInput


def equilateralTriangleS(a: float):
    if a < 0:
        raise ValueError("Require positive values")
    return a**2 * (sqrt(3) / 4)


edge_c = floatInput("Enter edge length:")
print(f"equilateralTriangleS({edge_c}) =", equilateralTriangleS(edge_c))

equilateralTriangleS(2.0) = 1.7320508075688772


# Ex9

- Run Ex7 with:

```
a: 2
b: 2
c: 2
A: 1.0471975512
B: 1.0471975512
C: 1.0471975512
```

result: 1.7320508075722796

- Run Ex8 with edge = 2:
  result: 1.7320508075722796

-> The results are equal.


In [26]:
# Ex10

from utils.formatInput import floatInput as fI

width, height = fI("Enter width: "), fI("Enter height: ")

if width <= 0 or height <= 0:
    raise ValueError("Invalid size value")

print(f"S_room(width={width}, height={height}) =", width * height)

S_room(width=10.0, height=10.0) = 100.0


In [27]:
# Ex11

from utils.formatInput import floatInput as fI

SQUARE_METER_TO_ACRE = 43560

width, height = fI("Enter width (Unit: m): "), fI("Enter height (Unit: m): ")

if width <= 0 or height <= 0:
    raise ValueError("Invalid size value")

print(f"S_field(width={width}, height={height}) =", width * height / 43560, "(Acre)")

S_field(width=100.0, height=220.0) = 0.5050505050505051 (Acre)


In [28]:
# Ex12

from utils.formatInput import floatInput

TIP_PERCENT = 18 / 100

# Không rõ yêu cầu đề, giả sử tiền thuế tính lên cả tiền tip
TAX_PERCENT = (1 + TIP_PERCENT) * 5 / 100

cost = floatInput("Enter meal cost:")
if cost <= 0:
    raise ValueError("Invalid cost value")

tax = round(cost * TAX_PERCENT, 2)
tip = round(cost * TIP_PERCENT, 2)

print("Food cost:", cost)
print("-\tTax:", tax)
print("-\tTip:", tip)
print("-\tTotal meal cost:", round(cost + tip + tax, 2))

Food cost: 10.0
-	Tax: 0.59
-	Tip: 1.8
-	Total meal cost: 12.39


In [ ]:
# Ex13

from utils.formatInput import intInput

decimal = intInput("Enter n:")
if decimal <= 0:
    raise ValueError("Invalid N+ value")

print(f"Sum(1 to {decimal}) =", int((decimal * (decimal + 1)) / 2))

Sum(1 to 10) = 55


In [30]:
# Ex14

from math import log10
from utils.formatInput import intInput

a, b = intInput("Enter integer a: "), intInput("Enter integer b: ")

print(f"a = {a}, b = {b}")
print("- Sum =", a + b)
print("- Difference =", a - b)
print("- Product =", a * b)

# div/mod 0 not defined
if b != 0:
    print("- Division =", a / b)
    print("- Remainder =", a % b)
else:
    print("- Division and remainder not defined (b is zero)")

# log10(a) valid when a > 0
if a > 0:
    print("- log10 =", log10(a))
else:
    print("- log10(a) not defined (a <= 0)")

print("- Power =", a**b)

a = 25, b = 4
- Sum = 29
- Difference = 21
- Product = 100
- Division = 6.25
- Remainder = 1
- log10 = 1.3979400086720377
- Power = 390625


In [31]:
# Ex15

from utils.formatInput import floatInput

C_WATER = 4.186
COST_PER_KWH = 8.9
JOULES_TO_KWH = 2.777e-7


def calc_Q(m: float, c: float, deltaT: float):
    result = m * c * deltaT
    if result < 0:
        raise Exception("Invalid input")
    return result


def calc_cost(Q: float):
    return Q * JOULES_TO_KWH * COST_PER_KWH


m, deltaT = floatInput("Enter m (g): "), floatInput("Enter deltaT (Celsius): ")

Q = calc_Q(m, C_WATER, deltaT)
print("Q (J) =", round(Q, 3))
print("Cost (cent) =", round(calc_cost(Q), 3))

Q (J) = 627900.0
Cost (cent) = 1.552


In [32]:
# Ex16

from utils.formatInput import floatInput
from math import sqrt

G = 9.8  # m/s^2
V_ZERO = 0  # m/s


def calc_free_fall_v(h: float, v_zero: float = V_ZERO):
    return sqrt(v_zero**2 + 2 * G * h)


print(
    "Speed when hits the ground (m/s):",
    calc_free_fall_v(floatInput("Enter height (m):")),
)

Speed when hits the ground (m/s): 44.271887242357316


In [33]:
# Ex17

from utils.formatInput import floatInput


def calc_WCI(temp_c: float, wind_speed_kmh: float) -> int:
    wci = (
        13.12
        + 0.6215 * temp_c
        - 11.37 * (wind_speed_kmh**0.16)
        + 0.3965 * temp_c * (wind_speed_kmh**0.16)
    )
    return round(wci)


# Nhập dữ liệu
Ta = floatInput("Enter air temperature (Celsius): ")
V = floatInput("Enter wind speed V (km/h): ")

# Tính WCI
WCI = calc_WCI(Ta, V)

print(f"Wind Chill Index (Ta={Ta}°C, V={V} km/h) = {WCI}")

Wind Chill Index (Ta=30.0°C, V=40.0 km/h) = 33


## Chapter 2 - Part 2


In [34]:
# Ex1

number = input(" Enter your positive integer: ")

if int(number) % 2 == 0:
    print(" This is an even number!")
else:
    print(" This is an odd number!")

 This is an odd number!


In [35]:
# Ex2

st = input("Enter your string: ")

if st.isupper():
    print("This is an upper string")
elif st.islower():
    print("This is an lower string")
else:
    print(" This string contains upper and lower characters ")

 This string contains upper and lower characters 


In [36]:
# Ex3

st = input("Enter your string: ")
st_search = input(" Enter a searching string :")

if st_search in st:
    print(" A search string was found, at location:", str(st.find(st_search)))
else:
    print("Not Found")

 A search string was found, at location: 3


In [37]:
# Ex4

from utils.formatInput import floatInput

num1 = floatInput("Enter first number: ")
num2 = floatInput("Enter second number: ")
num3 = floatInput("Enter third number: ")

if (num1 >= num2) and (num1 >= num3):
    largest = num1
elif (num2 >= num1) and (num2 >= num3):
    largest = num2
else:
    largest = num3

print("The largest number between", num1, ",", num2, "and", num3, "is", largest)

The largest number between 1.0 , -80.0 and -1.0 is 1.0


In [ ]:
# Ex5
from utils.formatInput import intInput

decimal = intInput(" Enter an integer greater than 0 : ")
assert decimal > 0, "Invalid input"

sum = 0.0

for num in range(1, decimal + 1):
    sum += float(float(num) / (num + 1))

print(sum)

7.980122655122655


In [39]:
# Ex6
st = input("Enter your sentence: ")

Sum_upper_Char = 0
Sum_lower_Char = 0

for c in st:
    if c.isupper():
        Sum_upper_Char += 1
    elif c.islower():
        Sum_lower_Char += 1
    else:
        pass

print("Sum of Upper Character:", Sum_upper_Char)
print("Sum of Lower Character:", Sum_lower_Char)

Sum of Upper Character: 3
Sum of Lower Character: 4


In [41]:
# Ex7
from utils.formatInput import intInput

num = intInput("Enter a number: ")
level = intInput("Bậc: ")
sum = 0
temp = num

while temp > 0:
    digit = temp % 10
    sum += digit**level
    temp //= 10

if num == sum:
    print(num, "is Amstrong, level : " + str(level))
else:
    print(num, "is not Amstrong")

153 is Amstrong, level : 3


In [42]:
# Ex 8

# Định nghĩa chuỗi các ký tự cần loại bỏ
st1 = """!()-[]{};:""\,<>./?@#$%^&*_~"""
my_str = "Hello!!!, he said ---and went{."

# Loại bỏ ký tự cần loại bỏ
st2 = ""
for char in my_str:
    if char not in st1:
        st2 = st2 + char

# Hiển thị chuỗi sau khi loại bỏ
print(st2)

Hello he said and went


In [43]:
# Ex9

my_str = "Hello this Is an Example With cased letters"

# Tách từ trong chuỗi và lưu vào danh sách ds_tu
ds_tu = my_str.split()

# Sắp xếp các phần tử (từ) trong danh sách từ ds_tu
ds_tu.sort()

# Hiển thị từ từ trong danh sách
print("Các từ đã được tách và sắp xếp theo Alphabe")
for tu in ds_tu:
    print(tu)

Các từ đã được tách và sắp xếp theo Alphabe
Example
Hello
Is
With
an
cased
letters
this


In [44]:
# Ex10

total = 0
count = 0

while True:
    inp = input("Enter a number: ")
    if inp == "":
        break
    value = float(inp)
    total = total + value
    count = count + 1

average = total / count
print("Average is:", average)

Average is: 3.0


In [45]:
# Ex11

numlist = list()


def sum(list: list):
    sum = 0
    for num in list:
        sum += num
    return sum


while True:
    inp = input("Enter a number: ")
    if inp == "":
        break
    value = float(inp)
    numlist.append(value)

average = sum(numlist) / len(numlist)
print("Giá trị Trung bình:", average)
print(numlist)

Giá trị Trung bình: 3.0
[1.0, 2.0, 3.0, 4.0, 5.0]


In [46]:
# Ex12

j = []

for num in range(2000, 3201):
    if (num % 7 == 0) and (num % 5 != 0):
        j.append(str(num))

print(",".join(j))

2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996,3003,3017,3024,3031,3038,3052,3059,3066,3073,3087,3094,3101,3108,3122,3129,3136,3143,3157,3164,3171,3178,3192,3199


In [47]:
# Ex13

lines = []

while True:
    s = input()
    if s:
        lines.append(s.upper())
    else:
        break

for sentence in lines:
    print(sentence)

HELLO,
WORLD


In [48]:
# Ex 14

founds = []

for i in range(100, 301):
    s = str(i)
    if all(int(digit) % 2 == 0 for digit in s):
        founds.append(s)

print(",".join(founds))

200,202,204,206,208,220,222,224,226,228,240,242,244,246,248,260,262,264,266,268,280,282,284,286,288


In [49]:
# Ex15

chuoi = input("Nhập vào các giá trị:")

kieu_ds = chuoi.split(",")
kieu_tuples = tuple(kieu_ds)

print(kieu_ds)
print(kieu_tuples)

['1', '2', '3']
('1', '2', '3')


In [ ]:
# Ex16

decimal = int(input("Nhập vào một số:"))
d = dict()

for i in range(1, decimal + 1):
    d[i] = i * i

print(d)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}


In [ ]:
# Ex17

chuoi = input("Nhập vào một chuỗi:")
d = {"chu_cai": 0, "chu_so": 0}

for ch in chuoi:
    if ch.isalpha() == True:
        d["chu_cai"] += 1
    elif ch.isdigit() == True:
        d["chu_so"] += 1
    else:
        pass

print("Số chữ cái là:", d["chu_cai"])
print("Số chữ số là:", d["chu_so"])

Số chữ cái là: 0
Số chữ số là: 2


In [53]:
# Ex 18


chuoi = input("Nhập vào một chuỗi:")
d = {"chu_hoa": 0, "chu_thuong": 0}

for ch in chuoi:
    if ch.isupper() == True:
        d["chu_hoa"] += 1
    elif ch.islower() == True:
        d["chu_thuong"] += 1
    else:
        pass

print("Số chữ hoa là:", d["chu_hoa"])
print("Số chữ thường là:", d["chu_thuong"])

Số chữ hoa là: 2
Số chữ thường là: 4


In [54]:
# Ex 19

ftext = open("note_mail_nop_bai.txt")
tu_dien_cac_tu = {}
for dong in ftext:
    danh_sach_tu = dong.split()
    for tu in danh_sach_tu:
        tu_dien_cac_tu[tu] = tu_dien_cac_tu.get(tu, 0) + 1

danh_sach = []
for key, val in tu_dien_cac_tu.items():
    newtup = (val, key)
    danh_sach.append(newtup)

danh_sach = sorted(danh_sach, reverse=True)

for val, key in danh_sach[:10]:
    print(key, val)

for 2
read 1
pythonvku001@gmail.com 1
file 1
exercise 1
19 1


In [55]:
# Ex20
from utils.formatInput import intInput

x = intInput("Nhập số cần tính giai thừa:")


def giaithua(x):
    return 1 if x == 0 else x * giaithua(x - 1)


print("Giai thừa của", x, "là:", giaithua(x))

Giai thừa của 5 là: 120


In [56]:
# Ex21


def Tao_In_DS():
    ds = list()
    for i in range(1, 21):
        ds.append(i**2)
    print(ds)


Tao_In_DS()

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


In [57]:
# Ex22


def create_and_print_list():
    lst = list()
    for i in range(1, 21):
        lst.append(i**2)
    print(lst[:5] + lst[(len(lst) - 5) :])


create_and_print_list()

[1, 4, 9, 16, 25, 256, 289, 324, 361, 400]


In [ ]:
# Ex23

from c2p2_modules.Kiem_tra import checkEven
from utils.formatInput import intInput

decimal = intInput("Enter a number to check:")
checkEven(decimal)

6 is even


In [59]:
# Ex24

from c2p2_modules.So_Sanh_Chuoi import printValue

st1, st2 = input("Nhập chuỗi 1:"), input("Nhập chuỗi 2:")
printValue(st1, st2)

Chuỗi dài hơn: hahahah


In [2]:
# Ex25

from c2p2_modules.tu_dien import create_dict, print_item, print_key, print_value

target_dict = create_dict(5)
print_item(target_dict)
print_key(target_dict)
print_value(target_dict)

Created dict(max=5): {1: 1, 2: 4, 3: 9, 4: 16, 5: 25}
Dict items:
	- Dict item: [1: 1]
	- Dict item: [2: 4]
	- Dict item: [3: 9]
	- Dict item: [4: 16]
	- Dict item: [5: 25]
Dict keys:
	- Dict key: [1]
	- Dict key: [2]
	- Dict key: [3]
	- Dict key: [4]
	- Dict key: [5]
Dict values:
	- Dict value: [1]
	- Dict value: [4]
	- Dict value: [9]
	- Dict value: [16]
	- Dict value: [25]


In [5]:
# Ex26

char = input("Enter char:")

vowels_exclude_y = {"a", "e", "i", "o", "u"}
if char == "y":
    print(char, "is vowel of consonant")
else:
    print(char, "is", "vowel" if char in vowels_exclude_y else "consonant")

x is consonant


In [9]:
# Ex27
from utils.formatInput import intInput

edges = intInput("Number of edges")
name_map = {
    3: "tam",
    4: "tứ",
    5: "ngũ",
    6: "lục",
    7: "thất",
    8: "bát",
    9: "cửu",
    10: "thập",
}
print((name_map[edges] + " giác").title() if (edges in name_map) else "Không hỗ trợ")

Ngũ Giác


In [13]:
# Ex28

month = intInput("Enter month:")
assert month > 0 and month < 13, "Invalid month"

if month == 2:
    print("28 or 29 days")
else:
    print("31" if (month in {1, 3, 5, 7, 8, 10, 12}) else "30", "days")

30 days


In [ ]:
# Ex29

a, b, c = (
    floatInput("Enter edge a"),
    floatInput("Enter edge b"),
    floatInput("Enter edge c"),
)

if a == b and a == c:
    print("Đều")
elif a == b or a == c or b == c:
    print("Cân")
else:
    print("Vuông, tù hoặc nhọn")

In [19]:
# Ex 30

year = intInput("Enter year:")


def nhuan(year: int):
    if year % 400 == 0:
        return True
    else:
        if year % 100 == 0:
            return False
        else:
            return True if year % 4 == 0 else False


print(year, ("" if nhuan(year) else "không phải ") + "là năm nhuận")

100 không phải là năm nhuận


In [20]:
# Ex31

shift = 3
plain_text = input("Enter message: ")
decoded = ""

for ch in plain_text:
    if "A" <= ch <= "Z":
        # Dịch trong phạm vi chữ hoa
        decoded += chr((ord(ch) - ord("A") + shift) % 26 + ord("A"))
    elif "a" <= ch <= "z":
        # Dịch trong phạm vi chữ thường
        decoded += chr((ord(ch) - ord("a") + shift) % 26 + ord("a"))
    else:
        # Giữ nguyên ký tự không phải chữ cái
        decoded += ch

print("Encoded message:", decoded)

Encoded message: Khoor, zruog


In [21]:
# Ex32

text = input("Enter message: ")
shift = intInput("Enter shift")


def ceasar_encode(plain_text, shift):
    encoded = ""
    for ch in plain_text:
        if "A" <= ch <= "Z":
            # Dịch trong phạm vi chữ hoa
            encoded += chr((ord(ch) - ord("A") + shift) % 26 + ord("A"))
        elif "a" <= ch <= "z":
            # Dịch trong phạm vi chữ thường
            encoded += chr((ord(ch) - ord("a") + shift) % 26 + ord("a"))
        else:
            # Giữ nguyên ký tự không phải chữ cái
            encoded += ch
    return encoded


def ceasar_decode(cipher_text, shift):
    decoded = ""
    for ch in cipher_text:
        if "A" <= ch <= "Z":
            # Dịch trong phạm vi chữ hoa
            decoded += chr((ord(ch) - ord("A") - shift) % 26 + ord("A"))
        elif "a" <= ch <= "z":
            # Dịch trong phạm vi chữ thường
            decoded += chr((ord(ch) - ord("a") - shift) % 26 + ord("a"))
        else:
            # Giữ nguyên ký tự không phải chữ cái
            decoded += ch
    return decoded


encoded = ceasar_encode(plain_text=text, shift=shift)
decoded = ceasar_decode(cipher_text=encoded, shift=shift)
print("Encoded message:", encoded)
print("Decoded message:", decoded)

Encoded message: Khoor, zruog
Decoded message: Hello, world


In [35]:
# Ex33

string = input("Enter string:")


def isPalindrom(string: str):
    length = len(string) // 2
    return all(string[i] == string[-(i + 1)] for i in range(0, length))


print(string, "is" + ("" if isPalindrom(string) else " not"), "Palindrom string")

hahah is Palindrom string


In [39]:
# Ex34


def toBinary(decimal: int):
    if decimal == 0:
        return "0"

    bits = ""
    while decimal > 0:
        bits = str(decimal % 2) + bits
        decimal //= 2
    return bits


print("Binary:", toBinary(intInput("Enter decimal")))

Binary: 101


In [43]:
# Ex35

i_list = input("Enter list of number:").split(",")
odds = []
for num in i_list:
    num = int(num)
    if num % 2 == 1:
        odds.append(num)

print("Odds:", odds)

Odds: [1, 3, 5]


In [8]:
# Ex36
from utils.formatInput import intInput

lst = []
while num := intInput("Enter number"):
    if num == 0:
        break
    lst.append(num)

lst.sort()
for num in lst:
    print(num)

1
2
3


In [10]:
# Ex37

lst = []
while line := input(""):
    lst.append(line)

print("\n".join(set(lst)))

Hello
world


In [18]:
# Ex38

neg = []
zer = []
pos = []

while num := input("Enter number"):
    numVal = int(num)
    if numVal == 0:
        zer.append(num)
    else:
        (neg if numVal < 0 else pos).append(num)

neg.extend(zer)
neg.extend(pos)

print(", ".join(neg))

-4, -1, -2, 0, 0, 3, 1


In [19]:
# Ex39

dct = {}
for i in range(1, intInput("Enter n") + 1):
    dct[i] = i**2
print(dct)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}


In [23]:
# Ex40

t = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
l = len(t) // 2
print(t[:l], t[l:])

(1, 2, 3, 4, 5) (6, 7, 8, 9, 10)


In [1]:
# Ex41

t = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
even_lst = []

for num in t:
    if num % 2 == 0:
        even_lst.append(num)

print(tuple(even_lst))

(2, 4, 6, 8, 10)


In [2]:
# Ex42


def gt(num: int):
    gt = 1
    for i in range(1, num + 1):
        gt *= i
    return gt


print(gt(8))

40320


In [7]:
# Ex43


def print_str(st1: str, st2: str):
    d_len = len(st1) - len(st2)
    if d_len == 0:
        return print(st1, "\n" + st2)
    print(st1 if d_len > 0 else st2)
    
print_str('ngan', 'dai')

ngan


In [8]:
# Ex44


def print_p2_lst(n: int = 20):
    lst = []
    for i in range(1, n + 1):
        lst.append(i**2)
    print(lst)
    
print_p2_lst()

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


In [9]:
# Ex45


def print_5_first_p2_lst(n: int = 20):
    lst = []
    for i in range(1, n + 1):
        lst.append(i**2)
    print(lst[:5])


print_5_first_p2_lst()

[1, 4, 9, 16, 25]


In [12]:
# Ex46


def print_5_last_p2_lst(n: int = 20):
    lst = []
    for i in range(1, n + 1):
        lst.append(i**2)
    print(lst[(len(lst) - 5) :])


print_5_last_p2_lst()

[256, 289, 324, 361, 400]


In [14]:
# Ex47

def print_ignore_5_first_p2_lst(n: int = 20):
    lst = []
    for i in range(1, n + 1):
        lst.append(i**2)
    print(lst[5:])


print_ignore_5_first_p2_lst()

[36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


In [16]:
# Ex48
from utils.formatInput import floatInput


def avg_3(a: float, b: float, c: float):
    return (a + b + c) / 3


a, b, c = floatInput("Enter a"), floatInput("Enter b"), floatInput("Enter c")
print(avg_3(a, b, c))

2.0


In [18]:
# Ex49

from math import sqrt
from utils.formatInput import intInput


def is_prime(n: int) -> bool:
    if n < 2:
        return False

    sqrtN = int(sqrt(n))
    for i in range(2, sqrtN + 1):
        if n % i == 0:
            return False
    return True


print("Is prime" if is_prime(intInput("Enter n")) else "Is not prime")

Is not prime


In [19]:
# Ex50

import random


def random_pw():
    return "".join(chr(random.randint(33, 126)) for _ in range(random.randint(7, 10)))


print(random_pw())

ubX[4.e5I


In [20]:
# Ex51


def check_pw(password):
    return (
        len(password) >= 8
        and any(c.isupper() for c in password)
        and any(c.islower() for c in password)
        and any(c.isdigit() for c in password)
    )


# Chương trình chính
pw = input("Enter password: ")
print("Good" if check_pw(pw) else "Bad")

Bad


In [21]:
# Ex52


def is_perfect(n):
    if n < 2:
        return False
    divisors_sum = sum(i for i in range(1, n // 2 + 1) if n % i == 0)
    return divisors_sum == n


print("Perfect numbers between 1 and 10000:")
for number in range(1, 10001):
    if is_perfect(number):
        print(number)

Perfect numbers between 1 and 10000:
6
28
496
8128


In [22]:
# Ex53


def all_sublists(lst):
    subs = [[]]
    for i in range(len(lst) + 1):
        for j in range(i + 1, len(lst) + 1):
            subs.append(lst[i:j])
    return subs


print(all_sublists([1, 2, 3]))

[[], [1], [1, 2], [1, 2, 3], [2], [2, 3], [3]]


In [23]:
# Ex54


def format_list(words):
    if not words:
        return ""
    if len(words) == 1:
        return words[0]
    if len(words) == 2:
        return f"{words[0]} and {words[1]}"
    return f"{', '.join(words[:-1])} and {words[-1]}"


# Main program
items = input("Enter words separated by spaces: ").split()
message = format_list(items)
print("Formatted list:", message)

Formatted list: apples, oranges, bananas and lemons


In [ ]:
# Ex55

